metricflow/plan_conversion/dataflow_to_sql.py

```diff
@@ -2019,6 +2020,29 @@ def strip_time_from_dt(ts: dt.datetime) -> dt.datetime:
         )
 
 
+@dataclass(frozen=True)
+class CteGenerationResult:
+    """This stores parameters for creating a dataset from a CTE."""
+
+    source_dataflow_plan_node_id: NodeId
+    cte_node: SqlCteNode
+    select_columns: Tuple[SqlSelectColumn, ...]
```

Comment:
Wondering if this attribute is redundant since you could get the select columns from the `SqlCteNode`. Might require making `SqlCteNode.select_statement` a `SqlSelectStatementNode` instead of a generic `SqlPlanNode` so we get the type benefits. Not sure if there are any limitations that prevent that, though.

Reply:
So the columns from the CTE can't be used in that way right now because the select statement in the CTE might have columns that are expressions. e.g.

```
WITH cte_0 AS (
  SELECT col_0 + col_1 AS col_2
  ...
)
```

I had plans to fix the signature of `SqlDataSet` to make it possible to use the `SqlCteNode`directly. The `SqlDataSet` initializer currently looks like

```
    def __init__(
        self,
        instance_set: InstanceSet,
        sql_select_node: Optional[SqlSelectStatementNode] = None,
        sql_node: Optional[SqlPlanNode] = None,
    ) -> None:
```

 which is a little odd. Once that's addressed, the SELECT could be eliminated.

Reply:
Sounds good! Not crucial for this PR.